### PR TITLE
Add new device type for BLE serial ports

### LINKED ISSUE #149: Reconnect to BLE serial interface fails after Android Bluetooth restart
Connecting a BLE serial interface device to XCSoar works, and the device will automatically be reconnected while that device's information remains in the Android Bluetooth cache. If the Bluetooth  system is restarted (e.g. turn Bluetooth off and on) or the phone is rebooted, the systems's cached information  about the device is lost. Then when BluetoothHelper#connect() is called with the MAC address of the device, it is assumed to be a legacy (RFCOMM) device and a connection attempt is made using legacy Bluetooth calls. This fails.

A side effect of this is that the device name is also lost, so it is displayed with the MAC address only in the device list.

So two changes are needed:

1) record the device type (BLE vs RFCOMM)
2) cache the device name for display purposes

Item 1) could be achieved by adding a new device type in DeviceConfig::PortType and changing the interface to the Android code. This won't solve 2)

An alternative would be to encode the device type and name into the MAC address string. This would require a change to the way the mac address is stored in the Config object.

Another alternative is to maintain persistent state for the device in the Android specific code, for example using a SharedPreferences instance or a file in the XCSoar directory. This would map each MAC address to a device type and name. This would not require any change to the non-Android code. The SharedPreferences option is easy, but has the drawback that copying the XCSoar default data directory to a new device would not transfer that information. Creating a file in the XCSoar directory would require knowledge of where that is, and a cursory inspection did not reveal any existing mechanism for Android Java code to access that in XCSoar.

My plan is to implement the last alternative using a SharedPreferences instance. Comments?

### PR DESCRIPTION
**Pull request by clydebarrow**
Brief summary of the changes
----------------------------
Add a new device type to represent Bluetooth Low Energy serial port devices. These must be distinguished from legacy Bluetooth serial devices using the RFCOMM profile since the connection sequence is different. The old code queried the Android BT system to try to determine the type of a device based solely on its address. This fails if the Android BT cache has been flushed (by device or BT system restart.) This commit adds a new device type - BLE_HM10 - representing a BLE device implementing a serial port on a specific characteristic.

The stored type is used to select the correct re-connection protocol.

Related issues and discussions
------------------------------

Fixes #149

### CODE REVIEW

**Comment by MaxKellermann** [comment]
From a first glance, this looks good! I have some minor complaints, but first this one:
Why "Catch exception and return null rather than throwing"? By propagating exceptions, the caller can catch and obtain error details, e.g. to show it to the user.
Propagating errors and showing details to the user is something XCSoar is currently bad at. I've tried to improve 7.0 a lot, but we're not there, existing code is mostly bad.
But can we make your new code better, as an example of how to do it right?

**Comment by clydebarrow** [comment]
I can see that you have changed the way in which exceptions are handled in V7, but for 6.8 I followed the same pattern as the existing code. 

I have done a similar patch for v7 with exception propogation, but there are issues: e.g. v7 has added a new port type "GLIDER_LINK" but it is incompletely implemented, which makes adding yet another port type problematic. I think I have sorted that out, and I can build v7 but it does not run well in several regards so I have not been able to properly test it.

**Comment by MaxKellermann** [comment]
Following the same pattern as existing code is not a good idea, when existing code is known to be bad.
I don't understand your Glider Link problem. What issue number are you talking about?

**Comment by clydebarrow** [comment]
> Following the same pattern as existing code is not a good idea, when existing code is known to be bad.

Well, the code in 6.8 is what it is, bad or not. I could not rewrite the whole exception handling in that branch - I had to make it work with what is there.

> I don't understand your Glider Link problem. What issue number are you talking about?

There is no issue raised AFAIK, but when `GLIDER_LINK` was added as an enum constant in `src/Device/Config.hpp` there should have been a corresponding entry added to the `port_type_strings` array in `src/Profile/DeviceConfig.cpp`. That broke my new code until I added that entry. Right now that's a moot problem since 7.0 is unusable anyway.

**Comment by MaxKellermann** [comment]
> That broke my new code until I added that entry.

I don't see your PR for adding this missing entry.

> 7.0 is unusable anyway.

Which bug number is that?

File: src/Dialogs/Device/DeviceEditWidget.cpp [hunk 1]
```diff
@@ -236,37 +236,51 @@ FillAndroidBluetoothPorts(DataFieldEnum &df, const DeviceConfig &config)
...
       ? env->GetStringUTFChars(name, NULL)
       : NULL;
 
-    AddPort(df, DeviceConfig::PortType::RFCOMM, address2, name2);
+    jstring devType = (jstring)env->GetObjectArrayElement(bonded, i * BLUETOOTH_LIST_STRIDE + 2);
+    const char *devType2 = devType != NULL
+      ? env->GetStringUTFChars(devType, NULL)
+      : NULL;
+
+    const DeviceConfig::PortType portType = strcmp("BLE",devType2) == 0 ? DeviceConfig::PortType::BLE_HM10 : DeviceConfig::PortType::RFCOMM;
```
Comment (MaxKellermann) [inline src/Dialogs/Device/DeviceEditWidget.cpp:268]: As codacy correctly identified: this is a nullptr dereference crash bug.

**Comment by MaxKellermann** [comment]
f64d9063434163407247f9e4964831a2458b694d

**Pull request closed without merging**